### PR TITLE
Pulled together Dot/BracketMember state transition logic and stack logic...

### PR DESCRIPTION
**Pull request by inverno**
... into the MemberProgramPoint.

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/se/points/MemberProgramPoint.java [hunk 1]
```diff
@@ -38,13 +43,53 @@ public static boolean originatesFrom(Tree element) {
   }
 
   @Override
-  public Optional<ProgramState> execute(ProgramState state) {
-    SymbolicValue symbolicValue = getObjectSymbolicValue(state);
-    return state.constrain(symbolicValue, Constraint.NOT_NULLY);
+  public Optional<ProgramState> execute(final ProgramState state) {
+    final SymbolicValue objectValue;
+    final ExpressionStack expressionStack = state.getStack();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/points/MemberProgramPoint.java]: MINOR Remove this use of "getStack"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/points/MemberProgramPoint.java [hunk 2]
```diff
@@ -38,13 +43,53 @@ public static boolean originatesFrom(Tree element) {
...
+    final ExpressionStack newExpressionStack;
+    Optional<ProgramState> newState;
+    if (element.is(Tree.Kind.BRACKET_MEMBER_EXPRESSION)) {
+      objectValue = state.peekStack(1);
+      newState = state.constrain(objectValue, Constraint.NOT_NULLY);
+      if(!newState.isPresent()) {
+        return newState;
+      }
+      newExpressionStack = expressionStack.apply(stack -> {
+        stack.pop(); // popping the array index
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/points/MemberProgramPoint.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ExpressionStack.java#L255](https://github.com/SonarSource/sonar-javascript/blob/f49da54c47c2c7d0012ec46fcb4e7a8f3f50997e/javascript-frontend/src/main/java/org/sonar/javascript/se/ExpressionStack.java#L255): Remove this unused private "executeDotMemberExpression" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MINOR][MINOR] [ExpressionStack.java#L196](https://github.com/SonarSource/sonar-javascript/blob/f49da54c47c2c7d0012ec46fcb4e7a8f3f50997e/javascript-frontend/src/main/java/org/sonar/javascript/se/ExpressionStack.java#L196): This case's code block is the same as the block for the case on line 109. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1871)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: javascript-frontend/src/test/java/org/sonar/javascript/se/points/MemberProgramPointTest.java [hunk 3]
```diff
@@ -86,4 +103,11 @@ public void matchingTrees() throws Exception {
     assertThat(MemberProgramPoint.originatesFrom(tree(Tree.Kind.FOR_OF_STATEMENT))).isFalse();
     assertThat(MemberProgramPoint.originatesFrom(tree(Tree.Kind.AND_ASSIGNMENT))).isFalse();
   }
+
+  @Test
+  public void fillStackWithResolvedDotMemberValue() throws Exception {
+    state = state.pushToStack(BuiltInObjectSymbolicValue.STRING_PROTOTYPE);
+    assertThat(dotMemberPoint.execute(state).isPresent()).isTrue();
+    dotMemberPoint.execute(state).ifPresent(s -> assertThat(s.getConstraint(s.peekStack())).isEqualTo(Constraint.POSITIVE_NUMBER_PRIMITIVE.or(Constraint.ZERO)));
```
Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/se/points/MemberProgramPointTest.java]: @inverno Can we replace `ifPresent` with `get`?

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Pull request merged**